// jupyterlab/lib/commandpalette/plugin.js
/*-----------------------------------------------------------------------------
| Copyright (c) Jupyter Development Team.
| Distributed under the terms of the Modified BSD License.
|----------------------------------------------------------------------------*/
"use strict";
var disposable_1 = require("@phosphor/disposable");
var widgets_1 = require("@phosphor/widgets");
var instancerestorer_1 = require("../instancerestorer");
var _1 = require("./");
/**
 * A thin wrapper around the `CommandPalette` class to conform with the
 * JupyterLab interface for the application-wide command palette.
 */
var Palette = (function () {
    /**
     * Create a palette instance.
     */
    function Palette(palette) {
        this._palette = null;
        this._palette = palette;
    }
    Object.defineProperty(Palette.prototype, "placeholder", {
        get: function () {
            return this._palette.inputNode.placeholder;
        },
        /**
         * The placeholder text of the command palette's search input.
         */
        set: function (placeholder) {
            this._palette.inputNode.placeholder = placeholder;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Activate the command palette for user input.
     */
    Palette.prototype.activate = function () {
        this._palette.activate();
    };
    /**
     * Add a command item to the command palette.
     *
     * @param options - The options for creating the command item.
     *
     * @returns A disposable that will remove the item from the palette.
     */
    Palette.prototype.addItem = function (options) {
        var _this = this;
        var item = this._palette.addItem(options);
        return new disposable_1.DisposableDelegate(function () { return _this._palette.removeItem(item); });
    };
    return Palette;
}());
/**
 * The default commmand palette extension.
 */
var plugin = {
    activate: activate,
    id: 'jupyter.services.commandpalette',
    provides: _1.ICommandPalette,
    requires: [instancerestorer_1.IInstanceRestorer],
    autoStart: true
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Export the plugin as default.
 */
exports.default = plugin;
/**
 * Activate the command palette.
 */
function activate(app, restorer) {
    var commands = app.commands;
    var palette = new widgets_1.CommandPalette({ commands: commands });
    // Let the application restorer track the command palette for restoration of
    // application state (e.g. setting the command palette as the current side bar
    // widget).
    restorer.add(palette, 'command-palette');
    palette.id = 'command-palette';
    palette.title.label = 'Commands';
    app.commands.addCommand(_1.CommandIDs.activate, {
        execute: function () { app.shell.activateLeft(palette.id); },
        label: 'Activate Command Palette'
    });
    app.commands.addCommand(_1.CommandIDs.hide, {
        execute: function () {
            if (!palette.isHidden) {
                app.shell.collapseLeft();
            }
        },
        label: 'Hide Command Palette'
    });
    app.commands.addCommand(_1.CommandIDs.toggle, {
        execute: function () {
            if (palette.isHidden) {
                return app.commands.execute(_1.CommandIDs.activate, void 0);
            }
            return app.commands.execute(_1.CommandIDs.hide, void 0);
        },
        label: 'Toggle Command Palette'
    });
    palette.inputNode.placeholder = 'SEARCH';
    app.shell.addToLeftArea(palette);
    return new Palette(palette);
}
