// jupyterlab/lib/csvwidget/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var instancetracker_1 = require("../common/instancetracker");
var docregistry_1 = require("../docregistry");
var filebrowser_1 = require("../filebrowser");
var instancerestorer_1 = require("../instancerestorer");
var widget_1 = require("./widget");
/**
 * The name of the factory that creates CSV widgets.
 */
var FACTORY = 'Table';
/**
 * The table file handler extension.
 */
var plugin = {
    activate: activate,
    id: 'jupyter.extensions.csv-handler',
    requires: [docregistry_1.IDocumentRegistry, instancerestorer_1.IInstanceRestorer],
    autoStart: true
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Export the plugin as default.
 */
exports.default = plugin;
/**
 * Activate the table widget extension.
 */
function activate(app, registry, restorer) {
    var factory = new widget_1.CSVWidgetFactory({
        name: FACTORY,
        fileExtensions: ['.csv'],
        defaultFor: ['.csv']
    });
    var tracker = new instancetracker_1.InstanceTracker({ namespace: 'csvwidget' });
    // Handle state restoration.
    restorer.restore(tracker, {
        command: filebrowser_1.CommandIDs.open,
        args: function (widget) { return ({ path: widget.context.path, factory: FACTORY }); },
        name: function (widget) { return widget.context.path; }
    });
    registry.addWidgetFactory(factory);
    factory.widgetCreated.connect(function (sender, widget) {
        // Track the widget.
        tracker.add(widget);
        // Notify the instance tracker if restore data needs to update.
        widget.context.pathChanged.connect(function () { tracker.save(widget); });
    });
}
