// jupyterlab/lib/csvwidget/table.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var dsv = require("d3-dsv");
var signaling_1 = require("@phosphor/signaling");
var virtualdom_1 = require("@phosphor/virtualdom");
var vdom_1 = require("../common/vdom");
var widget_1 = require("../renderers/widget");
/**
 * The hard limit on the number of rows to display.
 */
exports.DISPLAY_LIMIT = 1000;
/**
 * The class name added to a csv table widget.
 */
var CSV_TABLE_CLASS = 'jp-CSVTable';
/**
 * A CSV table content model.
 */
var CSVModel = (function (_super) {
    __extends(CSVModel, _super);
    /**
     * Instantiate a CSV model.
     */
    function CSVModel(options) {
        if (options === void 0) { options = {}; }
        var _this = _super.call(this) || this;
        _this._maxExceeded = new signaling_1.Signal(_this);
        _this._content = options.content || '';
        _this._delimiter = options.delimiter || ',';
        return _this;
    }
    Object.defineProperty(CSVModel.prototype, "maxExceeded", {
        /**
         * A signal emitted when the parsed value's rows exceed the display limit. It
         * emits the length of the parsed value.
         */
        get: function () {
            return this._maxExceeded;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CSVModel.prototype, "content", {
        /**
         * The raw model content.
         */
        get: function () {
            return this._content;
        },
        set: function (content) {
            if (this._content === content) {
                return;
            }
            this._content = content;
            this.triggerChange();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CSVModel.prototype, "delimiter", {
        /**
         * The CSV delimiter value.
         */
        get: function () {
            return this._delimiter;
        },
        set: function (delimiter) {
            if (this._delimiter === delimiter) {
                return;
            }
            this._delimiter = delimiter;
            this.triggerChange();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Parse the content using the model's delimiter.
     *
     * #### Notes
     * This method will always return parsed content that has at most the display
     * limit worth of rows, currently maxing out at 1000 rows.
     */
    CSVModel.prototype.parse = function () {
        var output = dsv.dsvFormat(this._delimiter).parse(this._content);
        var available = output.length;
        var maximum = exports.DISPLAY_LIMIT;
        if (available > maximum) {
            // Mutate the array instead of slicing in order to conserve memory.
            output.splice(maximum);
            this._maxExceeded.emit({ available: available, maximum: maximum });
        }
        return output;
    };
    return CSVModel;
}(vdom_1.VDomModel));
exports.CSVModel = CSVModel;
/**
 * A CSV table content widget.
 */
var CSVTable = (function (_super) {
    __extends(CSVTable, _super);
    /**
     * Instantiate a new CSV table widget.
     */
    function CSVTable() {
        var _this = _super.call(this) || this;
        _this.addClass(CSV_TABLE_CLASS);
        _this.addClass(widget_1.HTML_COMMON_CLASS);
        return _this;
    }
    /**
     * Render the content as virtual DOM nodes.
     */
    CSVTable.prototype.render = function () {
        if (!this.model) {
            return virtualdom_1.h.table([virtualdom_1.h.thead(), virtualdom_1.h.tbody()]);
        }
        var rows = this.model.parse();
        var cols = rows.columns || [];
        return virtualdom_1.h.table([
            virtualdom_1.h.thead(cols.map(function (col) { return virtualdom_1.h.th(col); })),
            virtualdom_1.h.tbody(rows.map(function (row) { return virtualdom_1.h.tr(cols.map(function (col) { return virtualdom_1.h.td(row[col]); })); }))
        ]);
    };
    return CSVTable;
}(vdom_1.VDomWidget));
exports.CSVTable = CSVTable;
