// jupyterlab/lib/codeeditor/widget.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var widgets_1 = require("@phosphor/widgets");
/**
 * A widget which hosts a code editor.
 */
var CodeEditorWidget = (function (_super) {
    __extends(CodeEditorWidget, _super);
    /**
     * Construct a new code editor widget.
     */
    function CodeEditorWidget(options) {
        var _this = _super.call(this) || this;
        _this._editor = null;
        _this._needsRefresh = false;
        _this._editor = options.factory({
            host: _this.node,
            model: options.model,
            uuid: options.uuid,
            wordWrap: options.wordWrap,
            readOnly: options.readOnly,
            selectionStyle: options.selectionStyle
        });
        return _this;
    }
    Object.defineProperty(CodeEditorWidget.prototype, "editor", {
        /**
         * Get the editor wrapped by the widget.
         */
        get: function () {
            return this._editor;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CodeEditorWidget.prototype, "model", {
        /**
         * Get the model used by the widget.
         */
        get: function () {
            return this._editor.model;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Dispose of the resources held by the widget.
     */
    CodeEditorWidget.prototype.dispose = function () {
        if (this.isDisposed) {
            return;
        }
        _super.prototype.dispose.call(this);
        this._editor.dispose();
        this._editor = null;
    };
    /**
     * Handle `'activate-request'` messages.
     */
    CodeEditorWidget.prototype.onActivateRequest = function (msg) {
        this._editor.focus();
    };
    /**
     * A message handler invoked on an `'after-attach'` message.
     */
    CodeEditorWidget.prototype.onAfterAttach = function (msg) {
        _super.prototype.onAfterAttach.call(this, msg);
        this.node.addEventListener('focus', this, true);
        if (this.isVisible) {
            this._editor.refresh();
            this._needsRefresh = false;
        }
    };
    /**
     * Handle `before_detach` messages for the widget.
     */
    CodeEditorWidget.prototype.onBeforeDetach = function (msg) {
        this.node.removeEventListener('focus', this, true);
    };
    /**
     * A message handler invoked on an `'after-show'` message.
     */
    CodeEditorWidget.prototype.onAfterShow = function (msg) {
        this._editor.refresh();
        this._needsRefresh = false;
    };
    /**
     * A message handler invoked on an `'resize'` message.
     */
    CodeEditorWidget.prototype.onResize = function (msg) {
        if (msg.width >= 0 && msg.height >= 0) {
            this._editor.setSize(msg);
            this._needsRefresh = false;
        }
        else if (this._editor.hasFocus()) {
            this._editor.refresh();
            this._needsRefresh = false;
        }
        else {
            this._needsRefresh = true;
        }
    };
    /**
     * Handle the DOM events for the widget.
     *
     * @param event - The DOM event sent to the widget.
     *
     * #### Notes
     * This method implements the DOM `EventListener` interface and is
     * called in response to events on the panel's DOM node. It should
     * not be called directly by user code.
     */
    CodeEditorWidget.prototype.handleEvent = function (event) {
        switch (event.type) {
            case 'focus':
                this._evtFocus(event);
                break;
            default:
                break;
        }
    };
    /**
     * Handle `focus` events for the widget.
     */
    CodeEditorWidget.prototype._evtFocus = function (event) {
        if (this._needsRefresh) {
            this._editor.refresh();
            this._needsRefresh = false;
        }
    };
    return CodeEditorWidget;
}(widgets_1.Widget));
exports.CodeEditorWidget = CodeEditorWidget;
