// jupyterlab/lib/about/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var commandpalette_1 = require("../commandpalette");
var instancetracker_1 = require("../common/instancetracker");
var instancerestorer_1 = require("../instancerestorer");
var _1 = require("./");
/**
 * The about page extension.
 */
var plugin = {
    activate: activate,
    id: 'jupyter.extensions.about',
    autoStart: true,
    requires: [commandpalette_1.ICommandPalette, instancerestorer_1.IInstanceRestorer]
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Export the plugin as default.
 */
exports.default = plugin;
function activate(app, palette, restorer) {
    var namespace = 'about-jupyterlab';
    var model = new _1.AboutModel({ version: app.info.version });
    var command = _1.CommandIDs.open;
    var category = 'Help';
    var tracker = new instancetracker_1.InstanceTracker({ namespace: namespace });
    restorer.restore(tracker, {
        command: command,
        args: function () { return null; },
        name: function () { return 'about'; }
    });
    var widget;
    function newWidget() {
        var widget = new _1.AboutWidget();
        widget.model = model;
        widget.id = 'about';
        widget.title.label = 'About';
        widget.title.closable = true;
        tracker.add(widget);
        return widget;
    }
    app.commands.addCommand(command, {
        label: 'About JupyterLab',
        execute: function () {
            if (!widget || widget.isDisposed) {
                widget = newWidget();
                app.shell.addToMainArea(widget);
            }
            app.shell.activateMain(widget.id);
        }
    });
    palette.addItem({ command: command, category: category });
}
