// jupyterlab/lib/codemirror/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
require("codemirror/addon/edit/matchbrackets.js");
require("codemirror/addon/edit/closebrackets.js");
require("codemirror/addon/comment/comment.js");
require("codemirror/keymap/vim.js");
var widgets_1 = require("@phosphor/widgets");
var codeeditor_1 = require("../codeeditor");
var commandpalette_1 = require("../commandpalette");
var editorwidget_1 = require("../editorwidget");
var mainmenu_1 = require("../mainmenu");
var _1 = require(".");
/**
 * The editor services.
 */
exports.servicesPlugin = {
    id: codeeditor_1.IEditorServices.name,
    provides: codeeditor_1.IEditorServices,
    activate: function () { return _1.editorServices; }
};
/**
 * The editor commands.
 */
exports.commandsPlugin = {
    id: 'jupyter.services.editor-commands',
    requires: [editorwidget_1.IEditorTracker, mainmenu_1.IMainMenu, commandpalette_1.ICommandPalette],
    activate: activateEditorCommands,
    autoStart: true
};
/**
 * Export the plugins as default.
 */
var plugins = [exports.commandsPlugin, exports.servicesPlugin];
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = plugins;
/**
 * The command IDs used by the editor plugin.
 */
var CommandIDs;
(function (CommandIDs) {
    CommandIDs.lineNumbers = 'editor:line-numbers';
    CommandIDs.lineWrap = 'editor:line-wrap';
    CommandIDs.matchBrackets = 'editor:match-brackets';
    CommandIDs.vimMode = 'editor:vim-mode';
    CommandIDs.changeTheme = 'editor:change-theme';
    CommandIDs.createConsole = 'editor:create-console';
    CommandIDs.runCode = 'editor:run-code';
})(CommandIDs = exports.CommandIDs || (exports.CommandIDs = {}));
;
/**
 * Set up the editor widget menu and commands.
 */
function activateEditorCommands(app, tracker, mainMenu, palette) {
    var commands = app.commands;
    /**
     * Toggle editor matching brackets
     */
    function toggleMatchBrackets() {
        if (tracker.currentWidget) {
            var editor = tracker.currentWidget.editor;
            if (editor instanceof _1.CodeMirrorEditor) {
                var cm = editor.editor;
                cm.setOption('matchBrackets', !cm.getOption('matchBrackets'));
            }
        }
    }
    /**
     * Toggle the editor's vim mode
     */
    function toggleVim() {
        tracker.forEach(function (widget) {
            if (widget.editor instanceof _1.CodeMirrorEditor) {
                var cm = widget.editor.editor;
                var keymap = cm.getOption('keyMap') === 'vim' ? 'default'
                    : 'vim';
                cm.setOption('keyMap', keymap);
            }
        });
    }
    /**
     * Create a menu for the editor.
     */
    function createMenu() {
        var settings = new widgets_1.Menu({ commands: commands });
        var theme = new widgets_1.Menu({ commands: commands });
        var menu = new widgets_1.Menu({ commands: commands });
        menu.title.label = 'Editor';
        settings.title.label = 'Settings';
        theme.title.label = 'Theme';
        settings.addItem({ command: CommandIDs.lineNumbers });
        settings.addItem({ command: CommandIDs.lineWrap });
        settings.addItem({ command: CommandIDs.matchBrackets });
        settings.addItem({ command: CommandIDs.vimMode });
        commands.addCommand(CommandIDs.changeTheme, {
            label: function (args) { return args['theme']; },
            execute: function (args) {
                var name = args['theme'] || _1.CodeMirrorEditor.DEFAULT_THEME;
                tracker.forEach(function (widget) {
                    if (widget.editor instanceof _1.CodeMirrorEditor) {
                        var cm = widget.editor.editor;
                        cm.setOption('theme', name);
                    }
                });
            }
        });
        [
            'jupyter', 'default', 'abcdef', 'base16-dark', 'base16-light',
            'hopscotch', 'material', 'mbo', 'mdn-like', 'seti', 'the-matrix',
            'xq-light', 'zenburn'
        ].forEach(function (name) { return theme.addItem({
            command: 'editor:change-theme',
            args: { theme: name }
        }); });
        menu.addItem({ type: 'separator' });
        menu.addItem({ type: 'submenu', submenu: settings });
        menu.addItem({ type: 'submenu', submenu: theme });
        return menu;
    }
    mainMenu.addMenu(createMenu(), { rank: 30 });
    commands.addCommand(CommandIDs.matchBrackets, {
        execute: function () { toggleMatchBrackets(); },
        label: 'Toggle Match Brackets',
    });
    commands.addCommand(CommandIDs.vimMode, {
        execute: function () { toggleVim(); },
        label: 'Toggle Vim Mode'
    });
    [
        CommandIDs.lineNumbers,
        CommandIDs.lineWrap,
        CommandIDs.matchBrackets,
        CommandIDs.vimMode,
        CommandIDs.createConsole,
        CommandIDs.runCode
    ].forEach(function (command) { return palette.addItem({ command: command, category: 'Editor' }); });
}
