// jupyterlab/lib/markdownwidget/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var instancetracker_1 = require("../common/instancetracker");
var docregistry_1 = require("../docregistry");
var filebrowser_1 = require("../filebrowser");
var instancerestorer_1 = require("../instancerestorer");
var rendermime_1 = require("../rendermime");
var widget_1 = require("./widget");
/**
 * The class name for all main area portrait tab icons.
 */
var PORTRAIT_ICON_CLASS = 'jp-MainAreaPortraitIcon';
/**
 * The class name for the text editor icon from the default theme.
 */
var TEXTEDITOR_ICON_CLASS = 'jp-ImageTextEditor';
/**
 * The name of the factory that creates markdown widgets.
 */
var FACTORY = 'Rendered Markdown';
/**
 * The markdown handler extension.
 */
var plugin = {
    activate: activate,
    id: 'jupyter.extensions.rendered-markdown',
    requires: [docregistry_1.IDocumentRegistry, rendermime_1.IRenderMime, instancerestorer_1.IInstanceRestorer],
    autoStart: true
};
/**
 * Activate the markdown plugin.
 */
function activate(app, registry, rendermime, restorer) {
    var factory = new widget_1.MarkdownWidgetFactory({
        name: FACTORY,
        fileExtensions: ['.md'],
        rendermime: rendermime
    });
    var namespace = 'rendered-markdown';
    var tracker = new instancetracker_1.InstanceTracker({ namespace: namespace });
    var icon = PORTRAIT_ICON_CLASS + " " + TEXTEDITOR_ICON_CLASS;
    // Handle state restoration.
    restorer.restore(tracker, {
        command: filebrowser_1.CommandIDs.open,
        args: function (widget) { return ({ path: widget.context.path, factory: FACTORY }); },
        name: function (widget) { return widget.context.path; }
    });
    factory.widgetCreated.connect(function (sender, widget) {
        widget.title.icon = icon;
        // Notify the instance tracker if restore data needs to update.
        widget.context.pathChanged.connect(function () { tracker.save(widget); });
        tracker.add(widget);
    });
    registry.addWidgetFactory(factory);
}
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Export the plugin as default.
 */
exports.default = plugin;
