// jupyterlab/lib/rendermime/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var services_1 = require("@jupyterlab/services");
var commandlinker_1 = require("../commandlinker");
var filebrowser_1 = require("../filebrowser");
var _1 = require("./");
/**
 * The default rendermime provider.
 */
var plugin = {
    id: 'jupyter.services.rendermime',
    requires: [commandlinker_1.ICommandLinker],
    provides: _1.IRenderMime,
    activate: activate,
    autoStart: true
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Export the plugin as default.
 */
exports.default = plugin;
/**
 * Activate the rendermine plugin.
 */
function activate(app, linker) {
    var linkHandler = {
        handleLink: function (node, path) {
            if (!services_1.utils.urlParse(path).protocol && path.indexOf('//') !== 0) {
                linker.connectNode(node, filebrowser_1.CommandIDs.open, { path: path });
            }
        }
    };
    var items = _1.RenderMime.getDefaultItems();
    return new _1.RenderMime({ items: items, linkHandler: linkHandler });
}
;
