// jupyterlab/lib/completer/handler.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var messaging_1 = require("@phosphor/messaging");
/**
 * A class added to editors that can host a completer.
 */
exports.COMPLETABLE_CLASS = 'jp-mod-completable';
/**
 * A class added to editors that have an active completer.
 */
exports.COMPLETER_ACTIVE_CLASS = 'jp-mod-completer-active';
/**
 * A completion handler for editors.
 */
var CompletionHandler = (function () {
    /**
     * Construct a new completion handler for a widget.
     */
    function CompletionHandler(options) {
        this._editor = null;
        this._completer = null;
        this._kernel = null;
        this._pending = 0;
        this._completer = options.completer;
        this._completer.selected.connect(this.onCompletionSelected, this);
        this._completer.visibilityChanged.connect(this.onVisibilityChanged, this);
        this._kernel = options.kernel || null;
    }
    Object.defineProperty(CompletionHandler.prototype, "completer", {
        /**
         * The completer widget managed by the handler.
         */
        get: function () {
            return this._completer;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CompletionHandler.prototype, "editor", {
        /**
         * The editor used by the completion handler.
         */
        get: function () {
            return this._editor;
        },
        set: function (newValue) {
            if (newValue === this._editor) {
                return;
            }
            var editor = this._editor;
            // Clean up and disconnect from old editor.
            if (editor && !editor.isDisposed) {
                var model = editor.model;
                editor.host.classList.remove(exports.COMPLETABLE_CLASS);
                model.selections.changed.disconnect(this.onSelectionsChanged, this);
                model.value.changed.disconnect(this.onTextChanged, this);
            }
            // Reset completer state.
            this._completer.reset();
            // Update the editor and signal connections.
            editor = this._editor = newValue;
            if (editor) {
                var model = editor.model;
                model.selections.changed.connect(this.onSelectionsChanged, this);
                model.value.changed.connect(this.onTextChanged, this);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CompletionHandler.prototype, "isDisposed", {
        /**
         * Get whether the completion handler is disposed.
         */
        get: function () {
            return this._completer === null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CompletionHandler.prototype, "kernel", {
        /**
         * The kernel used by the completion handler.
         */
        get: function () {
            return this._kernel;
        },
        set: function (value) {
            this._kernel = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Dispose of the resources used by the handler.
     */
    CompletionHandler.prototype.dispose = function () {
        this._completer = null;
        this._kernel = null;
        // Use public accessor to disconnect from editor signals.
        this.editor = null;
    };
    /**
     * Invoke the handler and launch a completer.
     */
    CompletionHandler.prototype.invoke = function () {
        messaging_1.MessageLoop.sendMessage(this, CompletionHandler.Msg.InvokeRequest);
    };
    /**
     * Process a message sent to the completion handler.
     */
    CompletionHandler.prototype.processMessage = function (msg) {
        switch (msg.type) {
            case CompletionHandler.Msg.InvokeRequest.type:
                this.onInvokeRequest(msg);
                break;
            default:
                break;
        }
    };
    /**
     * Get the state of the text editor at the given position.
     */
    CompletionHandler.prototype.getState = function (position) {
        var editor = this.editor;
        var coords = editor.getCoordinateForPosition(position);
        return {
            text: editor.model.value.text,
            lineHeight: editor.lineHeight,
            charWidth: editor.charWidth,
            coords: coords,
            line: position.line,
            column: position.column
        };
    };
    /**
     * Make a complete request using the kernel.
     */
    CompletionHandler.prototype.makeRequest = function (position) {
        var _this = this;
        if (!this._kernel) {
            return Promise.reject(new Error('no kernel for completion request'));
        }
        var editor = this.editor;
        if (!editor) {
            return Promise.reject(new Error('No active editor'));
        }
        var offset = editor.getOffsetAt(position);
        var content = {
            code: editor.model.value.text,
            cursor_pos: offset
        };
        var pending = ++this._pending;
        var request = this.getState(position);
        return this._kernel.requestComplete(content).then(function (msg) {
            if (_this.isDisposed) {
                return;
            }
            // If a newer completion request has created a pending request, bail.
            if (pending !== _this._pending) {
                return;
            }
            _this.onReply(request, msg);
        });
    };
    /**
     * Handle `invoke-request` messages.
     */
    CompletionHandler.prototype.onInvokeRequest = function (msg) {
        // If there is neither a kernel nor a completer model, bail.
        if (!this._kernel || !this._completer.model) {
            return;
        }
        // If a completer session is already active, bail.
        if (this._completer.model.original) {
            return;
        }
        var editor = this._editor;
        this.makeRequest(editor.getCursorPosition());
    };
    /**
     * Receive a completion reply from the kernel.
     *
     * @param state - The state of the editor when completion request was made.
     *
     * @param reply - The API response returned for a completion request.
     */
    CompletionHandler.prototype.onReply = function (state, reply) {
        var model = this._completer.model;
        if (!model) {
            return;
        }
        // Completion request failures or negative results fail silently.
        var value = reply.content;
        if (value.status !== 'ok') {
            model.reset(true);
            return;
        }
        // Update the original request.
        model.original = state;
        // Update the options.
        model.setOptions(value.matches || []);
        // Update the cursor.
        model.cursor = { start: value.cursor_start, end: value.cursor_end };
    };
    /**
     * Handle selection changed signal from an editor.
     *
     * #### Notes
     * If a sub-class reimplements this method, then that class must either call
     * its super method or it must take responsibility for adding and removing
     * the completer completable class to the editor host node.
     */
    CompletionHandler.prototype.onSelectionsChanged = function () {
        var model = this._completer.model;
        if (!model) {
            return;
        }
        var editor = this._editor;
        var host = editor.host;
        var position = editor.getCursorPosition();
        var line = editor.getLine(position.line);
        if (line.match(/^\W*$/)) {
            model.reset(true);
            host.classList.remove(exports.COMPLETABLE_CLASS);
            return;
        }
        host.classList.add(exports.COMPLETABLE_CLASS);
    };
    /**
     * Handle a text changed signal from an editor.
     */
    CompletionHandler.prototype.onTextChanged = function () {
        var model = this._completer.model;
        if (!model) {
            return;
        }
        // If there is a text selection, no completion is allowed.
        var editor = this.editor;
        var _a = editor.getSelection(), start = _a.start, end = _a.end;
        if (start.column !== end.column || start.line !== end.line) {
            return;
        }
        // Allow completion if the current or a preceding char is not whitespace.
        var position = editor.getCursorPosition();
        var request = this.getState(position);
        this._completer.model.handleTextChange(request);
    };
    /**
     * Handle a visiblity change signal from a completion widget.
     */
    CompletionHandler.prototype.onVisibilityChanged = function (completion) {
        // Completer is not active.
        if (completion.isDisposed || completion.isHidden) {
            if (this._editor) {
                this._editor.host.classList.remove(exports.COMPLETER_ACTIVE_CLASS);
                this._editor.focus();
            }
            return;
        }
        // Completer is active.
        if (this._editor) {
            this._editor.host.classList.add(exports.COMPLETER_ACTIVE_CLASS);
        }
    };
    /**
     * Handle a completion selected signal from the completion widget.
     */
    CompletionHandler.prototype.onCompletionSelected = function (widget, value) {
        var model = this._completer.model;
        var editor = this._editor;
        if (!editor || !model) {
            return;
        }
        var patch = model.createPatch(value);
        if (!patch) {
            return;
        }
        var offset = patch.offset, text = patch.text;
        editor.model.value.text = text;
        var position = editor.getPositionAt(offset);
        editor.setCursorPosition(position);
    };
    return CompletionHandler;
}());
exports.CompletionHandler = CompletionHandler;
/**
 * A namespace for cell completion handler statics.
 */
(function (CompletionHandler) {
    /**
     * A namespace for completion handler messages.
     */
    var Msg;
    (function (Msg) {
        /* tslint:disable */
        /**
         * A singleton `'invoke-request'` message.
         */
        Msg.InvokeRequest = new messaging_1.Message('invoke-request');
        /* tslint:enable */
    })(Msg = CompletionHandler.Msg || (CompletionHandler.Msg = {}));
})(CompletionHandler = exports.CompletionHandler || (exports.CompletionHandler = {}));
exports.CompletionHandler = CompletionHandler;
