// jupyterlab/lib/common/observablejson.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var coreutils_1 = require("@phosphor/coreutils");
var messaging_1 = require("@phosphor/messaging");
var widgets_1 = require("@phosphor/widgets");
var virtualdom_1 = require("@phosphor/virtualdom");
var codeeditor_1 = require("../codeeditor");
var observablemap_1 = require("./observablemap");
/**
 * The class name added to a ObservableJSONWidget instance.
 */
var METADATA_CLASS = 'jp-ObservableJSONWidget';
/**
 * The class name added when the Metadata editor contains invalid JSON.
 */
var ERROR_CLASS = 'jp-mod-error';
/**
 * The class name added to the editor host node.
 */
var HOST_CLASS = 'jp-ObservableJSONWidget-host';
/**
 * The class name added to the button area.
 */
var BUTTON_AREA_CLASS = 'jp-ObservableJSONWidget-buttons';
/**
 * The class name added to the revert button.
 */
var REVERT_CLASS = 'jp-ObservableJSONWidget-revertButton';
/**
 * The class name added to the commit button.
 */
var COMMIT_CLASS = 'jp-ObservableJSONWidget-commitButton';
/**
 * A concrete Observable map for JSON data.
 */
var ObservableJSON = (function (_super) {
    __extends(ObservableJSON, _super);
    /**
     * Construct a new observable JSON object.
     */
    function ObservableJSON(options) {
        if (options === void 0) { options = {}; }
        return _super.call(this, {
            itemCmp: coreutils_1.JSONExt.deepEqual,
            values: options.values
        }) || this;
    }
    /**
     * Serialize the model to JSON.
     */
    ObservableJSON.prototype.toJSON = function () {
        var out = Object.create(null);
        for (var _i = 0, _a = this.keys(); _i < _a.length; _i++) {
            var key = _a[_i];
            var value = this.get(key);
            if (coreutils_1.JSONExt.isPrimitive(value)) {
                out[key] = value;
            }
            else {
                out[key] = JSON.parse(JSON.stringify(value));
            }
        }
        return out;
    };
    return ObservableJSON;
}(observablemap_1.ObservableMap));
exports.ObservableJSON = ObservableJSON;
/**
 * The namespace for ObservableJSON static data.
 */
(function (ObservableJSON) {
    /**
     * An observable JSON change message.
     */
    var ChangeMessage = (function (_super) {
        __extends(ChangeMessage, _super);
        /**
         * Create a new metadata changed message.
         */
        function ChangeMessage(args) {
            var _this = _super.call(this, 'jsonvalue-changed') || this;
            _this.args = args;
            return _this;
        }
        return ChangeMessage;
    }(messaging_1.Message));
    ObservableJSON.ChangeMessage = ChangeMessage;
})(ObservableJSON = exports.ObservableJSON || (exports.ObservableJSON = {}));
exports.ObservableJSON = ObservableJSON;
/**
 * A widget for editing observable JSON.
 */
var ObservableJSONWidget = (function (_super) {
    __extends(ObservableJSONWidget, _super);
    /**
     * Construct a new metadata editor.
     */
    function ObservableJSONWidget(options) {
        var _this = _super.call(this, { node: Private.createEditorNode() }) || this;
        _this._dataDirty = false;
        _this._inputDirty = false;
        _this._source = null;
        _this._originalValue = null;
        _this._changeGuard = false;
        _this.addClass(METADATA_CLASS);
        var host = _this.editorHostNode;
        var model = new codeeditor_1.CodeEditor.Model();
        model.value.text = 'No data!';
        model.mimeType = 'application/json';
        model.value.changed.connect(_this._onValueChanged, _this);
        _this.model = model;
        _this.editor = options.editorFactory({ host: host, model: model });
        _this.editor.readOnly = true;
        return _this;
    }
    Object.defineProperty(ObservableJSONWidget.prototype, "editorHostNode", {
        /**
         * Get the editor host node used by the metadata editor.
         */
        get: function () {
            return this.node.getElementsByClassName(HOST_CLASS)[0];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ObservableJSONWidget.prototype, "revertButtonNode", {
        /**
         * Get the revert button used by the metadata editor.
         */
        get: function () {
            return this.node.getElementsByClassName(REVERT_CLASS)[0];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ObservableJSONWidget.prototype, "commitButtonNode", {
        /**
         * Get the commit button used by the metadata editor.
         */
        get: function () {
            return this.node.getElementsByClassName(COMMIT_CLASS)[0];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ObservableJSONWidget.prototype, "source", {
        /**
         * The observable source.
         */
        get: function () {
            return this._source;
        },
        set: function (value) {
            if (this._source === value) {
                return;
            }
            if (this._source) {
                this._source.changed.disconnect(this._onSourceChanged, this);
            }
            this._source = value;
            this.editor.readOnly = !value;
            if (value) {
                value.changed.connect(this._onSourceChanged, this);
            }
            this._setValue();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ObservableJSONWidget.prototype, "isDirty", {
        /**
         * Get whether the editor is dirty.
         */
        get: function () {
            return this._dataDirty || this._inputDirty;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Handle the DOM events for the widget.
     *
     * @param event - The DOM event sent to the widget.
     *
     * #### Notes
     * This method implements the DOM `EventListener` interface and is
     * called in response to events on the notebook panel's node. It should
     * not be called directly by user code.
     */
    ObservableJSONWidget.prototype.handleEvent = function (event) {
        switch (event.type) {
            case 'blur':
                this._evtBlur(event);
                break;
            case 'click':
                this._evtClick(event);
                break;
            default:
                break;
        }
    };
    /**
     * Handle `after-attach` messages for the widget.
     */
    ObservableJSONWidget.prototype.onAfterAttach = function (msg) {
        var node = this.editorHostNode;
        node.addEventListener('blur', this, true);
        this.revertButtonNode.hidden = true;
        this.commitButtonNode.hidden = true;
        this.revertButtonNode.addEventListener('click', this);
        this.commitButtonNode.addEventListener('click', this);
    };
    /**
     * Handle `before_detach` messages for the widget.
     */
    ObservableJSONWidget.prototype.onBeforeDetach = function (msg) {
        var node = this.editorHostNode;
        node.removeEventListener('blur', this, true);
        this.revertButtonNode.removeEventListener('click', this);
        this.commitButtonNode.removeEventListener('click', this);
    };
    /**
     * Handle a change to the metadata of the source.
     */
    ObservableJSONWidget.prototype._onSourceChanged = function (sender, args) {
        if (this._changeGuard) {
            return;
        }
        if (this._inputDirty || this.editor.hasFocus()) {
            this._dataDirty = true;
            return;
        }
        this._setValue();
    };
    /**
     * Handle change events.
     */
    ObservableJSONWidget.prototype._onValueChanged = function () {
        var valid = true;
        try {
            var value = JSON.parse(this.editor.model.value.text);
            this.removeClass(ERROR_CLASS);
            this._inputDirty = (!this._changeGuard && !coreutils_1.JSONExt.deepEqual(value, this._originalValue));
        }
        catch (err) {
            this.addClass(ERROR_CLASS);
            this._inputDirty = true;
            valid = false;
        }
        this.revertButtonNode.hidden = !this._inputDirty;
        this.commitButtonNode.hidden = !valid || !this._inputDirty;
    };
    /**
     * Handle blur events for the text area.
     */
    ObservableJSONWidget.prototype._evtBlur = function (event) {
        // Update the metadata if necessary.
        if (!this._inputDirty && this._dataDirty) {
            this._setValue();
        }
    };
    /**
     * Handle click events for the buttons.
     */
    ObservableJSONWidget.prototype._evtClick = function (event) {
        var target = event.target;
        if (target === this.revertButtonNode) {
            this._setValue();
        }
        else if (target === this.commitButtonNode) {
            if (!this.commitButtonNode.hidden && !this.hasClass(ERROR_CLASS)) {
                this._changeGuard = true;
                this._mergeContent();
                this._changeGuard = false;
                this._setValue();
            }
        }
    };
    /**
     * Merge the user content.
     */
    ObservableJSONWidget.prototype._mergeContent = function () {
        var model = this.editor.model;
        var current = this._getContent() || {};
        var old = this._originalValue;
        var user = JSON.parse(model.value.text);
        var source = this.source;
        // If it is in user and has changed from old, set in current.
        for (var key in user) {
            if (!coreutils_1.JSONExt.deepEqual(user[key], old[key])) {
                current[key] = user[key];
            }
        }
        // If it was in old and is not in user, remove from current.
        for (var key in old) {
            if (!(key in user)) {
                delete current[key];
                source.delete(key);
            }
        }
        // Set the values.
        for (var key in current) {
            source.set(key, current[key]);
        }
    };
    /**
     * Get the metadata from the owner.
     */
    ObservableJSONWidget.prototype._getContent = function () {
        var source = this._source;
        if (!source) {
            return void 0;
        }
        var content = {};
        for (var _i = 0, _a = source.keys(); _i < _a.length; _i++) {
            var key = _a[_i];
            content[key] = source.get(key);
        }
        return content;
    };
    /**
     * Set the value given the owner contents.
     */
    ObservableJSONWidget.prototype._setValue = function () {
        this._dataDirty = false;
        this._inputDirty = false;
        this.revertButtonNode.hidden = true;
        this.commitButtonNode.hidden = true;
        this.removeClass(ERROR_CLASS);
        var model = this.editor.model;
        var content = this._getContent();
        this._changeGuard = true;
        if (content === void 0) {
            model.value.text = 'No data!';
            this._originalValue = null;
        }
        else {
            var value = JSON.stringify(content, null, 2);
            model.value.text = value;
            this._originalValue = content;
        }
        this.editor.refresh();
        this._changeGuard = false;
        this.commitButtonNode.hidden = true;
        this.revertButtonNode.hidden = true;
    };
    return ObservableJSONWidget;
}(widgets_1.Widget));
exports.ObservableJSONWidget = ObservableJSONWidget;
/**
 * The namespace for module private data.
 */
var Private;
(function (Private) {
    /**
     * Create the node for the EditorWdiget.
     */
    function createEditorNode() {
        var cancelTitle = 'Revert changes to data';
        var confirmTitle = 'Commit changes to data';
        return virtualdom_1.VirtualDOM.realize(virtualdom_1.h.div({ className: METADATA_CLASS }, virtualdom_1.h.div({ className: BUTTON_AREA_CLASS }, virtualdom_1.h.span({ className: REVERT_CLASS, title: cancelTitle }), virtualdom_1.h.span({ className: COMMIT_CLASS, title: confirmTitle })), virtualdom_1.h.div({ className: HOST_CLASS })));
    }
    Private.createEditorNode = createEditorNode;
})(Private || (Private = {}));
