// jupyterlab/lib/help/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var services_1 = require("@jupyterlab/services");
var widgets_1 = require("@phosphor/widgets");
var about_1 = require("../about");
var iframe_1 = require("../common/iframe");
var instancetracker_1 = require("../common/instancetracker");
var commandpalette_1 = require("../commandpalette");
var faq_1 = require("../faq");
var instancerestorer_1 = require("../instancerestorer");
var mainmenu_1 = require("../mainmenu");
var statedb_1 = require("../statedb");
var _1 = require("./");
/**
 * A flag denoting whether the application is loaded over HTTPS.
 */
var LAB_IS_SECURE = window.location.protocol === 'https:';
/**
 * The class name added to the help widget.
 */
var HELP_CLASS = 'jp-Help';
/**
 * A list of help resources.
 */
var RESOURCES = [
    {
        text: 'Scipy Lecture Notes',
        url: 'http://www.scipy-lectures.org/'
    },
    {
        text: 'Numpy Reference',
        url: 'https://docs.scipy.org/doc/numpy/reference/'
    },
    {
        text: 'Scipy Reference',
        url: 'https://docs.scipy.org/doc/scipy/reference/'
    },
    {
        text: 'Notebook Tutorial',
        url: 'https://nbviewer.jupyter.org/github/jupyter/notebook/' +
            'blob/master/docs/source/examples/Notebook/Notebook Basics.ipynb'
    },
    {
        text: 'Python Reference',
        url: 'https://docs.python.org/3.5/'
    },
    {
        text: 'IPython Reference',
        url: 'https://ipython.org/documentation.html?v=20160707164940'
    },
    {
        text: 'Matplotlib Reference',
        url: 'http://matplotlib.org/contents.html?v=20160707164940'
    },
    {
        text: 'SymPy Reference',
        url: 'http://docs.sympy.org/latest/index.html?v=20160707164940'
    },
    {
        text: 'Pandas Reference',
        url: 'http://pandas.pydata.org/pandas-docs/stable/?v=20160707164940'
    },
    {
        text: 'Markdown Reference',
        url: 'https://help.github.com/articles/' +
            'getting-started-with-writing-and-formatting-on-github/'
    }
];
RESOURCES.sort(function (a, b) {
    return a.text.localeCompare(b.text);
});
/**
 * The help handler extension.
 */
var plugin = {
    activate: activate,
    id: 'jupyter.extensions.help-handler',
    requires: [mainmenu_1.IMainMenu, commandpalette_1.ICommandPalette, instancerestorer_1.IInstanceRestorer],
    autoStart: true
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Export the plugin as default.
 */
exports.default = plugin;
/*
  * An IFrame the disposes itself when closed.
  *
  * This is needed to clear the state restoration db when IFrames are closed.
 */
var ClosableIFrame = (function (_super) {
    __extends(ClosableIFrame, _super);
    function ClosableIFrame() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Dispose of the IFrame when closing.
     */
    ClosableIFrame.prototype.onCloseRequest = function (msg) {
        this.dispose();
    };
    return ClosableIFrame;
}(iframe_1.IFrame));
/**
 * Activate the help handler extension.
 *
 * @param app - The phosphide application object.
 *
 * returns A promise that resolves when the extension is activated.
 */
function activate(app, mainMenu, palette, restorer) {
    var counter = 0;
    var category = 'Help';
    var namespace = 'help-doc';
    var command = _1.CommandIDs.open;
    var menu = createMenu();
    var tracker = new instancetracker_1.InstanceTracker({ namespace: namespace });
    // Handle state restoration.
    restorer.restore(tracker, {
        command: command,
        args: function (widget) { return ({ url: widget.url, text: widget.title.label }); },
        name: function (widget) { return widget.url; }
    });
    /**
     * Create a new ClosableIFrame widget.
     */
    function newClosableIFrame(url, text) {
        var iframe = new ClosableIFrame();
        iframe.addClass(HELP_CLASS);
        iframe.title.label = text;
        iframe.title.closable = true;
        iframe.id = namespace + "-" + ++counter;
        iframe.url = url;
        tracker.add(iframe);
        return iframe;
    }
    /**
     * Create a menu for the help plugin.
     */
    function createMenu() {
        var commands = app.commands;
        var menu = new widgets_1.Menu({ commands: commands });
        menu.title.label = category;
        menu.addItem({ command: about_1.CommandIDs.open });
        menu.addItem({ command: faq_1.CommandIDs.open });
        menu.addItem({ command: _1.CommandIDs.launchClassic });
        menu.addItem({ type: 'separator' });
        RESOURCES.forEach(function (args) { menu.addItem({ args: args, command: command }); });
        menu.addItem({ type: 'separator' });
        menu.addItem({ command: statedb_1.CommandIDs.clear });
        return menu;
    }
    app.commands.addCommand(command, {
        label: function (args) { return args['text']; },
        execute: function (args) {
            var url = args['url'];
            var text = args['text'];
            // If help resource will generate a mixed content error, load externally.
            if (LAB_IS_SECURE && services_1.utils.urlParse(url).protocol !== 'https:') {
                window.open(url);
                return;
            }
            var iframe = newClosableIFrame(url, text);
            app.shell.addToMainArea(iframe);
            app.shell.activateMain(iframe.id);
        }
    });
    app.commands.addCommand(_1.CommandIDs.launchClassic, {
        label: 'Launch Classic Notebook',
        execute: function () { window.open(services_1.utils.getBaseUrl() + 'tree'); }
    });
    RESOURCES.forEach(function (args) { palette.addItem({ args: args, command: command, category: category }); });
    palette.addItem({ command: statedb_1.CommandIDs.clear, category: category });
    palette.addItem({ command: _1.CommandIDs.launchClassic, category: category });
    mainMenu.addMenu(menu, {});
}
